src/hardware/system_timer.rs

```diff
@@ -93,24 +79,80 @@ impl rtic::Monotonic for SystemTimer {
                     // Note(unsafe): We are in a critical section, so it is safe to read the
                     // global variable.
                     return unsafe {
-                        ((OVERFLOWS << 16) + current_value) as i32
+                        Ok(Instant::new(
+                            ((OVERFLOWS << 16) + current_value) as u32,
+                        ))
                     };
                 }
             }
         })
     }
+}
 
+impl Monotonic for SystemTimer {
     /// Reset the timer count.
-    unsafe fn reset() {
+    unsafe fn reset(&mut self) {
         // Note: The timer must be safely configured in `SystemTimer::initialize()`.
         let regs = &*hal::device::TIM15::ptr();
 
         OVERFLOWS = 0;
         regs.cnt.reset();
     }
 
-    /// Get a timestamp correlating to zero time.
-    fn zero() -> i32 {
-        0
+    fn set_compare(&mut self, instant: &Instant<Self>) {
+        let regs = unsafe { &*hal::device::TIM15::ptr() };
+
+        // Note(cs): When operating on the timer registers, we must be atomic to prevent
+        // pre-emption.
+        cortex_m::interrupt::free(|_cs| {
```

Comment:
Same here. Can we push the CS further down to only cover the register accesses (or the `match`)?
I am unsure what atomicicity you mean here. If two tasks are racing to do `set_compare()` they would first need to check that they only ever decrease the `ccr` (make it fire earlier). This is an external constraint independent of the atomicity here but I can't find where they do that.
Now even then, if we let them race, then what atomicity is in danger? Retrieving/computing `now` is already atomic. The racing `ccr` writes, the `cc1ie` sets, and the `pend()` are all atomic and can be done in any order among the tasks with the same result.

Reply:
I actually think we want to explicitly ignore a potential race on `set_compare()` - it's RTIC's job to handle scheduling. If two `set_compare()`s override eachother in different contexts, I'd consider that an RTIC bug. I don't think we can actually protect against that using atomic access, so I'm unsure if we need the critical sections at all.

Reply:
Agreed. But this also means we just can't use `set_compare()` in `NetworkProcessor` (in the future).

Reply:
I don't think we ever want to use `set_compare()` - this is an API explicitly provided for the RTIC scheduling purposes. Definitely worth noting though - I'll add some docs.